FLT/Assumptions/Mazur.lean

```diff
@@ -0,0 +1,100 @@
+and invoke known partial results towards the Birch and Swinnerton-Dyer
+conjecture by Bump--Friedberg--Hoffstein and Murty--Murty; this simplification
+was crucial in Merel's 1995 extension of Mazur's results to elliptic curves
+over general number fields.
+
+Mazur's paper is
+
+Mazur, Barry (1977). "Modular curves and the Eisenstein ideal".
+Publications Mathématiques de l'IHÉS. 47 (1): 33–186. doi:10.1007/BF02684339.
+MR 0488287. S2CID 122609075.
+
+## Why do we need this?
+
+Given a counterexample `x^p+y^p=z^p` to Fermat's Last Theorem, with `p ≥ 5`
+a prime, the first step is to construct an elliptic curve `E` called the Frey
+curve. It was observed by Frey that the p-torsion `E[p]` of `E` was a
+Galois representation which was unramified away from `2` and `p`, and whose
+ramification at `2` and `p` was also highly controlled. One way for this
+to happen would be for the representation to be reducible (for example it
+could be the sum of the trivial character and the mod `p` cyclotomic character).
+However, Mazur's theorem rules this out; if the trivial representation were
+a subrepresentation of `E[p]` then `E` would have a point of order `p ≥ 5`
+and this, combined with the fact that the 2-torsion in the Frey curve
+has size 4, would give us a torsion subgroup of size `4p > 16`, contradicting
+Mazur's theorem.
+
+More generally, Mazur's theorem and the fact that `E` is semistable can be
+used to deduce that the Galois representation `E[p]` is irreducible, a hypothesis
+which is crucial for applying the modularity lifting results in the proof that
+we are formalizing in this project. Historically the irreducibility was used
+in a different way; Wiles proved that the Frey curve was modular by arguing
+at 3 and 5 rather than at p, and this was known to imply FLT by earlier work
+of Ribet which relied on `E[p]` being irreducible. Our proof will avoid Ribet's
+work completely by applying Wiles' techniques at `p` rather than at small primes.
+
+## References
+
+- [Barry Mazur, Modular curves and the Eisenstein idea][mazur_torsion]
+
+At the time of writing, the article is available here
+https://www.numdam.org/item/?id=PMIHES_1977__47__33_0
+
+The github tracking issue for this assumption is #477 on the FLT github repository.
+
+-/
+
+open scoped WeierstrassCurve.Affine -- E⟮ℚ⟯ notation
+
+-- Implementation note: The `ncard` function used in the axiom returns the junk
+-- value 0 if it is fed an infinite set, so the axiom as it stands actually
+-- says "the size of the torsion subgroup is either at most 16, or infinite".
+-- However it is well-known (and much much much easier than Mazur's theorem)
+-- that the torsion subgroup is finite, so the axiom as stated will suffice
+-- for the application to FLT.
+
+/-- Mazur's bound for the size of the torsion subgroup of an elliptic curve
+over the rationals . -/
+axiom Mazur_statement (E : WeierstrassCurve ℚ) [E.IsElliptic] :
+    {P : E⟮ℚ⟯ | ∃ n ≥ 1, n • P = 0}.ncard ≤ 16
```

Comment:
I suppose you could do this too:
```suggestion
axiom Mazur_statement (E : WeierstrassCurve ℚ) [E.IsElliptic] :
    (AddCommGroup.torsion E⟮ℚ⟯ : Set E⟮ℚ⟯).ncard ≤ 16
```
There might be more API for `AddCommGroup.torsion` than a set you define yourself.